mongo/with_transactions_test.go

```diff
@@ -259,6 +259,103 @@ func TestConvenientTransactions(t *testing.T) {
 		assert.True(t, ok, "expected context for abortTransaction to contain ctxKey")
 		assert.Equal(t, "foobar", ctxValue, "expected value for ctxKey to be 'world', got %s", ctxValue)
 	})
+	t.Run("commitTransaction timeout allows abortTransaction", func(t *testing.T) {
+		coll := db.Collection("test")
+		// Explicitly create the collection on server because implicit collection creation is not allowed in
+		// transactions for server versions <= 4.2.
+		err := db.RunCommand(bgCtx, bson.D{{"create", coll.Name()}}).Err()
+		assert.Nil(t, err, "error creating collection on server: %v\n", err)
+
+		session, err := client.StartSession()
+		assert.Nil(t, err, "StartSession error: %v", err)
+
+		defer func() {
+			session.EndSession(bgCtx)
+			_ = coll.Drop(bgCtx)
+		}()
+
+		err = WithSession(bgCtx, session, func(sessionContext SessionContext) error {
+			// Start transaction
+			err = session.StartTransaction()
+			assert.Nil(t, err, "StartTransaction error: %v", err)
+
+			// Insert a document
+			_, err := coll.InsertOne(sessionContext, bson.D{{"val", 17}})
+			assert.Nil(t, err, "InsertOne error: %v", err)
+
+			// Set a timeout of 0 for commitTransaction
+			commitTimeoutCtx, commitCancel := context.WithTimeout(sessionContext, 0)
+			defer commitCancel()
+
+			// commitTransaction results in context.DeadlineExceeded
+			commitErr := session.CommitTransaction(commitTimeoutCtx)
+			assert.True(t, IsTimeout(commitErr),
+				"expected timeout error error; got %v", commitErr)
+
+			// assert session state is not Committed
+			clientSession := session.(XSession).ClientSession()
+			assert.False(t, clientSession.TransactionCommitted(), "expected session state to not be Committed")
+
+			// abortTransaction without error
+			abortErr := session.AbortTransaction(context.Background())
+			assert.Nil(t, abortErr, "AbortTransaction error: %v", abortErr)
+
+			return nil
+		})
+	})
+	t.Run("commitTransaction timeout does not retry", func(t *testing.T) {
```

Comment:
I have a couple of issues with this test:

1. It relies on our test timeout being 10 minutes. If that were to change, this test might mask a bug unless we make a corresponding change here.
2. I don't think this is actually testing what you want. Even if you remove the `IsDeadlineExceededError` check in `CommitLoop`, the test still passes. I believe this is because the first `commitTransaction` command fails with a network timeout, which causes the connection to be closed. The loop continues but the second `CommitTransaction` call fails because connection checkout times out. Connection checkout errors do not have the `UnknownTransactionCommitResult` label, so the commit loop breaks out early.

The following is a proposal to change this test which I believe addresses both issues, but it relies on switching from `IsDeadlineExceededError` to `ctx.Err() != nil`, which I recommended in a previous comment.

```
withTransactionTimeout := 2 * time.Second
// create collection and defer coll.Drop() like you're already doing 
// start session and defer EndSession() like you're already doing
defer sess.EndSession(...)

defer func() {
    // run {killAllSessions: []}, ignoring CommandError with code errorInterrupted
}()

cancelCtx, cancel := context.WithCancel(bgCtx)
defer cancel()

callback := func() {
    _, _ = sess.WithTransaction(cancelCtx, func(sessCtx SessionContext) (interface{} , error) {
            _, err := coll.InsertOne(sessCtx, bson.M{"x": 1})
            assert.Nil(t, err, "InsertOne error: %v", err)
            cancel()
            return nil, nil
    })
}

assert.Soon(t, callback, 500*time.Millisecond)
```

Reply:
Overview of how the proposed test works:

1. The callback asserts that the `InsertOne` call succeeds to make sure the transaction is started server-side. The callback itself returns no error to make sure we enter the commit loop in `WithTransaction`.
2. After the `InsertOne` runs and is verified, the callback cancels the context. The driver optimistically checks for context timeout/cancellation before doing network operations, so the first `CommitTransaction` call to fail _client-side_, rather than causing a network timeout like the previous test does. The exact check is [here](https://github.com/mongodb/mongo-go-driver/blob/10fbddeef32c3b0516ab2e935933fedf8bc2fd04/x/mongo/driver/topology/connection.go#L288).
3. Because the `CommitLoop` checks if `ctx.Err() != nil`, it will break early.
4. If you were to remove the `ctx.Err() != nil` check, it would keep trying `CommitTransaction`, which would keep failing client-side as described in (2). This failure doesn't cause the connection to be invalidated, but the error comes from the `connection#WriteWireMessage` function, so it's considered a network error and gets the `UnknownTransactionCommitResult` label, which makes it retryable by the commit loop. After 500ms, the `assert.Soon` call would fail.
5. The test leaves an open transaction on the server, so it runs `killAllSessions` manually in a defer function. Note that the location of the defer is important because it has to be done before we attempt to drop the collection. This command is known to interrupt itself, so we ignore the error if the code is `errorInterrupted`.

Reply:
Thanks for the deep explanation and walkthrough! Those reasons for changing the test make sense to me; I now agree that it wasn't testing exactly the right situation (didn't understand the difference between a client-side error and a network error).

Used the new test that you recommended above. I did not defer a call to `killAllSessions` since that is already done at the top of the test suite [here](https://github.com/mongodb/mongo-go-driver/blob/master/mongo/with_transactions_test.go#L43-L61). Would I still need another deferred `killAllSessions` just for this test?

Reply:
The existing deferred function will run once the top-level function is exiting, which happens after all tests have executed. This is fine for now as this is the last test in the file, but if someone were to add another test below it, that could cause issues. Test order shouldn't matter at all, so it'd be best to defer it again if we know this test leaves a transaction open on the server.

Reply:
Sounds good! Added it in.